### PR TITLE
examples/kernel_sample: move a workqueue sample into kernel_sample

### PR DESCRIPTION
**Pull request by YiHyunJin**
When run kernel_sample, also workqueue to test.

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/kernel_sample/kernel_sample_main.c [hunk 1]
```diff
@@ -468,8 +468,13 @@ static int user_main(int argc, char *argv[])
 	!defined(CONFIG_DISABLE_SIGNALS)
 		printf("\nuser_main: vfork() test\n");
 		vfork_test();
+		check_test_memory_usage();
 #endif
 
+#ifdef CONFIG_SCHED_WORKQUEUE
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/kernel_sample_main.c]: in original workqueue Kconfig, there is a optional dependency : SCHED_WORKQUEUE || LIB_USRWORK
please check the dependency

File: apps/examples/kernel_sample/kernel_sample.h [hunk 2]
```diff
@@ -211,6 +211,11 @@ void priority_inheritance(void);
 int vfork_test(void);
 #endif
 
+/* workqueue.c **************************************************************/
+#if defined(CONFIG_SCHED_WORKQUEUE) || defined(LIB_USRWORK)
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/kernel_sample.h]: please change LIB_USRWORK -> CONFIG_LIB_USRWORK

File: apps/examples/kernel_sample/kernel_sample_main.c [hunk 3]
```diff
@@ -468,8 +468,13 @@ static int user_main(int argc, char *argv[])
 	!defined(CONFIG_DISABLE_SIGNALS)
 		printf("\nuser_main: vfork() test\n");
 		vfork_test();
+		check_test_memory_usage();
 #endif
 
+#if defined(CONFIG_SCHED_WORKQUEUE) || defined(LIB_USRWORK)
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/kernel_sample_main.c]: same as above

File: apps/examples/kernel_sample/workqueue.c [hunk 4]
```diff
@@ -106,5 +107,12 @@ int workqueue_main(int argc, char *argv[])
 
 	printf("test4 is queued and will excute it after 90 ticks\n");
 	work_queue(HPWORK, test_wq4, wq_test4, NULL, 90);
-	return 0;
+	
+	usleep(SECOND_USEC);
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/workqueue.c]: why do you use usleep with huge value?
how about using sleep?

File: apps/examples/kernel_sample/workqueue.c [hunk 5]
```diff
@@ -72,16 +75,14 @@ static void wq_test4(FAR void *arg)
...
  ****************************************************************************/
 
-#ifdef CONFIG_BUILD_KERNEL
-int main(int argc, FAR char *argv[])
-#else
-int workqueue_main(int argc, char *argv[])
-#endif
+void workqueue_test(void)
 {
+#if defined(CONFIG_SCHED_WORKQUEUE) || defined(LIB_USRWORK)
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/workqueue.c]: I think you'd better bind the CONFIG condition with whole function

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/kernel_sample/workqueue.c (hunk 4 above)
Comment (YiHyunJin) [reply apps/examples/kernel_sample/workqueue.c]: sleep(1)? yes!

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/kernel_sample/workqueue.c [hunk 6]
```diff
@@ -23,6 +23,7 @@
 #include <stdio.h>
 #include <tinyara/clock.h>
 #include <tinyara/wqueue.h>
+#include "kernel_sample.h"
```
Comment (sunghan-chang) [inline apps/examples/kernel_sample/workqueue.c]: Why do you include this header? What's purpose?

File: apps/examples/kernel_sample/workqueue.c [hunk 7]
```diff
@@ -72,16 +73,14 @@ static void wq_test4(FAR void *arg)
...
  ****************************************************************************/
 
-#ifdef CONFIG_BUILD_KERNEL
-int main(int argc, FAR char *argv[])
-#else
-int workqueue_main(int argc, char *argv[])
-#endif
+void workqueue_test(void)
 {
+#if defined(CONFIG_SCHED_WORKQUEUE) || defined(CONFIG_LIB_USRWORK)
```
Comment (sunghan-chang) [inline apps/examples/kernel_sample/workqueue.c]: In others, you use it as below.
```
#if xxx
workqueue_test
#endif
```
But here
```
workqueue_test
#if xxx
#endif
```
Use a same policy.

File: apps/examples/kernel_sample/workqueue.c [hunk 8]
```diff
@@ -72,16 +73,14 @@ static void wq_test4(FAR void *arg)
...
 	systime_t cur_time;
+
```
Comment (sunghan-chang) [inline apps/examples/kernel_sample/workqueue.c]: Why do you make a empty line here?

File: apps/examples/kernel_sample/workqueue.c [hunk 9]
```diff
@@ -106,5 +105,12 @@ int workqueue_main(int argc, char *argv[])
 
 	printf("test4 is queued and will excute it after 90 ticks\n");
 	work_queue(HPWORK, test_wq4, wq_test4, NULL, 90);
-	return 0;
+	
+	sleep(1);
```
Comment (sunghan-chang) [inline apps/examples/kernel_sample/workqueue.c]: For our new implementation, we don't need to use sleep. Let's remove it.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/kernel_sample/workqueue.c [hunk 10]
```diff
@@ -32,6 +35,8 @@
  * Private Data
  ****************************************************************************/
 
+uint64_t wq_order[4];
```
Comment (sunghan-chang) [inline apps/examples/kernel_sample/workqueue.c]: Why not systime_t?

File: apps/examples/kernel_sample/workqueue.c [hunk 11]
```diff
@@ -40,47 +45,52 @@ static void wq_test1(FAR void *arg)
 	systime_t cur_time = 0;
 	cur_time = clock_systimer();
 
-	printf("test1 is excuted at (%d) ticks\n", cur_time);
+	wq_order[1] = (uint64_t)cur_time;
```
Comment (sunghan-chang) [inline apps/examples/kernel_sample/workqueue.c]: We need to discuss this method. This is not a wide usage. If user modifies time value, this is definitely wrong.

**Comment by sunghan-chang** [comment]
@YiHyunJin @jeongchanKim Any update?

File: apps/examples/kernel_sample/workqueue.c [hunk 12]
```diff
@@ -32,6 +36,15 @@
  * Private Data
  ****************************************************************************/
 
+struct workqueue_context{
+	uint64_t	excuted_time;
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/workqueue.c]: why uint64_t? I think systime_t is better
and check the spell : executed

File: apps/examples/kernel_sample/workqueue.c [hunk 13]
```diff
@@ -92,19 +133,42 @@ int workqueue_main(int argc, char *argv[])
...
+	
+	sleep(1);
+
+	free(test_wq1);
+	free(test_wq2);
+	free(test_wq3);
+	free(test_wq4);
+
+	for (index = 0; index < 4; index++) {
+		printf("workqueue %d was excuted after %lld ticks", g_wc[index].number, g_wc[index].excuted_time - (uint64_t)cur_time);
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/workqueue.c]: please check the spell

File: apps/examples/kernel_sample/workqueue.c [hunk 14]
```diff
@@ -40,47 +53,75 @@ static void wq_test1(FAR void *arg)
...
 
-	printf("test2 is excuted at (%d) ticks\n", cur_time);
+	pthread_mutex_lock(&mutex_lock);
+
+	g_wc[count].excuted_time = cur_time;
+	g_wc[count++].number = 2;
+
+	pthread_mutex_unlock(&mutex_lock);
+
+
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/workqueue.c]: please remove unnecessary line

File: apps/examples/kernel_sample/workqueue.c [hunk 15]
```diff
@@ -32,6 +36,15 @@
  * Private Data
  ****************************************************************************/
 
+struct workqueue_context{
+	uint64_t	excuted_time;
+	int		number;
+};
+
+int count;
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/workqueue.c]: please make name detailed

File: apps/examples/kernel_sample/workqueue.c [hunk 16]
```diff
@@ -32,6 +36,15 @@
...
+struct workqueue_context g_wc[4];
+pthread_mutex_t mutex_lock;
```
Comment (jeongchanKim) [inline apps/examples/kernel_sample/workqueue.c]: same as above

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/kernel_sample/workqueue.c:39: ERROR: [SPC_M_KWD] missing space after struct definition

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/kernel_sample/workqueue.c [hunk 17]
```diff
@@ -32,56 +36,54 @@
  * Private Data
  ****************************************************************************/
 
+static systime_t g_cur_time;
```
Comment (sunghan-chang) [inline apps/examples/kernel_sample/workqueue.c]: Let's change this name to g_standard_time or g_start_time or something else.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Comment by seinfra** [comment]
Code Rule Check OK.

**Pull request merged**